### PR TITLE
Add documentation on debugging OpenCL applications with Pocl

### PR DESCRIPTION
**Pull request by franz**
GDB part needs improvement

### CODE REVIEW

File: doc/sphinx/source/debug.rst [hunk 1]
```diff
@@ -0,0 +1,236 @@
+Debugging OpenCL applications with Pocl
```
Comment (pjaaskel) [inline doc/sphinx/source/debug.rst:1]: Let's shorten it PoCL from now on.

File: doc/sphinx/source/debug.rst [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+========================================
+
+
+There are several ways to debug applications with Pocl,
```
Comment (pjaaskel) [inline doc/sphinx/source/debug.rst:5]: PoCL

File: doc/sphinx/source/debug.rst [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+Corresponding pocl CMake flags to enable them are:
+``ENABLE_ASAN, ENABLE_LSAN, ENABLE_UBSAN, ENABLE_TSAN.``
+
+See also "handling LLVM" below.
+
+Downsides:
+  * requires rebuilding both the application and pocl;
+  * the application and pocl's runtime code are compiled with sanitizer,
+    but ATM the kernels are not compiled with sanitizer
```
Comment (pjaaskel) [inline doc/sphinx/source/debug.rst:74]: at the moment

File: doc/sphinx/source/debug.rst [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+Upsides:
+  * much faster than Valgrind
```
Comment (pjaaskel) [inline doc/sphinx/source/debug.rst:77]: Move this sanitizer after the valgrind method if referring to it?

File: doc/sphinx/source/debug.rst [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+		#1 0x562f6f33e493 in main /tmp/examples/example1/example1.c:74
+		#2 0x7f738df0bb96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)
+
+
+
+Debugging with Valgrind
+-----------------------------------------------
+
+Downsides:
+  * extremely slow esp with computationally intensive kernels
```
Comment (pjaaskel) [inline doc/sphinx/source/debug.rst:119]: especially

File: doc/sphinx/source/debug.rst [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+  * ocassionally false positive reports
```
Comment (pjaaskel) [inline doc/sphinx/source/debug.rst:120]: occassionally

File: doc/sphinx/source/debug.rst [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+	==18726==      possibly lost: 53,773,316 bytes in 524,329 blocks
+	==18726==    still reachable: 411,350,622 bytes in 73,488 blocks
+	==18726==         suppressed: 0 bytes in 0 blocks
+
+
+Debugging with GDB
+-----------------------------------------------
+
+The setup is similar to Valgrind; build pocl with ``-DCMAKE_BUILD_TYPE=Debug``
+and export ``POCL_EXTRA_BUILD_FLAGS="-g -cl-opt-disable"`` before running gdb.
```
Comment (pjaaskel) [inline doc/sphinx/source/debug.rst:187]: ...or just add these flags to clbprogram options.

**Comment by franz** [comment]
OK, fixed pocl to PoCL etc, plus updated the GDB section with examples.

Stepping inside kernel in GDB is possible, though a little tricky; i've added an example for that too.

File: doc/sphinx/source/debug.rst [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+	Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
+	[New Thread 0x7fffedf36700 (LWP 11851)]
+	[Switching to Thread 0x7fffedf36700 (LWP 11851)]
+
+	Thread 2 "example" hit Breakpoint 1, work_group_scheduler (k=0x7fffe9001640, thread_data=0x5555557ad500)
+		at /tmp/pocl_source/lib/CL/devices/pthread/pthread_scheduler.c:307
+	307	          k->workgroup ((uint8_t*)arguments, (uint8_t*)&pc,
+
+
+Now it's about to launch the ``dot_product`` kernel for a single workgroup, so let's find the kernel source file::
```
Comment (pjaaskel) [inline doc/sphinx/source/debug.rst:306]: Phew. Quite a bit of hoops to go through. Why not just make the POCL_DEBUG dump the temp file name more explicitly so it can be used in gdb?

Comment (franz) [reply doc/sphinx/source/debug.rst:306]: We can do that, but it won't change the fact that you have to stop at `k->workgroup`  (or somewhere post-compilation-before-execution). So it would really only save one step, the `info functions kernel_name`. If you think it's worth it, i can add the code.

Comment (pjaaskel) [reply doc/sphinx/source/debug.rst:306]: Can we use the kernel function name to set the breakpoint (including the PoCL WG naming)?

Comment (franz) [reply doc/sphinx/source/debug.rst]: Yes, that works, and it makes the process significantly simpler. I've updated the docs

File: lib/CL/clReleaseContext.c [hunk 9]
```diff
@@ -67,7 +67,6 @@ POname(clReleaseContext)(cl_context context) CL_API_SUFFIX__VERSION_1_0
 
       /* see below on why we don't call uninit_devices here anymore */
       --cl_context_count;
-      POCL_ATOMIC_DEC (cl_context_count);
```
Comment (pjaaskel) [inline lib/CL/clReleaseContext.c:70]: Why did you remove the atomic one?

**Comment by pjaaskel** [comment]
Fails to build for LLVM 8.
```
FlattenGlobals.cc:117:58: error: ‘class llvm::Function’ has no member named ‘hasOptNone’; did you mean ‘hasSection’?
         if (pocl::Workgroup::isKernelToProcess(*f) && f->hasOptNone())
                                                          ^~~~~~~~~~
                                                          hasSection
```
